lgc/interface/lgc/Pipeline.h

```diff
@@ -698,10 +698,16 @@ class Pipeline {
   // with irLink(). This is a static method in Pipeline, as it does not need a Pipeline object, and can be used
   // in the front-end before a shader is associated with a pipeline.
   //
-  // @param func : Shader entry-point function
-  // @param stage : Shader stage
+  // @param func : Function to mark
+  // @param stage : Shader stage, or ShaderStageInvalid if none
   static void markShaderEntryPoint(llvm::Function *func, ShaderStage stage);
 
+  // Get a function's shader stage, as marked by markShaderEntryPoint.
```

Comment:
This isn't entirely correct, is it? getShaderStage will return a shader stage as set even if not marked by markShaderEntryPoint.

The distinction is actually important if we don't inline functions: you could have free functions that still "belong" to a shader stage but aren't entry points. So, they wouldn't have been marked by markShaderEntryPoint, but it probably makes sense to set their shader stage. And on that note, perhaps setShaderStage should be available to the frontend as well?

Cc @tsymalla

Reply:
I agree with the first point, in the case of `NoInline` non-entrypoint functions, which we could get by just marking them in `llpcSpirvLowerTranslator`, the comment is not entirely correct.

Making `setShaderStage` available to the frontend sounds good. When having free functions in the module, they are initially without shader stage, the assignment can only happen by examining the call instructions inside an entrypoint function. To do the assignment, all calls inside a entrypoint function need to be processed before entering the middle-end (inside `llpcSpirvLowerTranslator`), right after the entrypoint has been marked with the corresponding shader stage. This also needs to happen for nested `NoInline` functions. So, having a way to do the `NoInline` <-> ShaderStage mapping as early as possible could reduce trouble, for instance, when processing the calls to the function inside `PatchEntryPointMutate`, where we currently assume that only entry-point functions and declarations appear inside a module which would not be the case with `NoInline` functions, which would require juggling around with the shader stages for such functions.

Reply:
Right, we don't really support non-inlined funcs currently. If we do want to in the future, I would argue that it is not by a non-inlined function being "associated" with a shader. What if there is some IR transformation that commons up two identical functions, or mostly commons up two similar functions? It is then up to the middle-end to determine how a function needs to be compiled (wave size etc) based on where it is called from, and that might require duplicating it if it has ended up being called from two shaders with different requirements.

So I would say that a non-shader function would not be marked with a shader stage by the front-end.

Unless you're going to tell me that spir-v allows a function to access stuff that is local to its calling shader, such as the shader inputs. In which case I need to rethink...

Reply:
Apparently spir-v does allow a function to access shader inputs etc from the shader it is associated with.

Currently we always inline those, so it is not a problem. If we ever decide not to, I think that should be sorted out in the front-end (e.g. by modifying the IR to pass in extra args), and not left to PatchEntryPointMutate. So the front-end would only mark the shader stage on actual shaders, not functions associated with them.

On that basis, is this change ok?

Reply:
I think it's fine to hold off on exposing setShaderStage until we know exactly how we want to deal with this situation; i.e., whether the middle-end is responsible for untangling calls and duplicating functions if necessary, or if an earlier pass inserts an explicit "context" argument, or what.

Personally, I'd still tone down the comment and just say "get a function's shader stage". After all, `irLink` does set this for non-entry points if I'm reading the code right.

Reply:
Ah, you're right, `irLink` does do that. But we only use and test that for compute, and I'm not even sure we do that.

Yes, I'll tone down the comment.